pkg/kubernetes/watcher.go

```diff
@@ -225,42 +234,46 @@ func (w *ObjListWatcher) Stop() {
 	close(w.stopChannel)
 }
 
-func (w *ObjListWatcher) handleAdd(obj interface{}) {
+func (w *ObjListWatcher) handleAdd(obj interface{}) error {
+	var err error
 	switch w.ResourceKind {
 	case podResourceType:
 		pod, ok := obj.(*corev1.Pod)
 		if !ok {
-			klog.Infof("Could not convert obj: %v", w.ResourceKind)
-			return
+			return fmt.Errorf("could not convert obj: %v", w.ResourceKind)
 		}
 		for _, condition := range pod.Status.Conditions {
 			if condition.Type != corev1.ContainersReady {
+				// set the error in case we reach the end of the loop and no ContainersReady condition is found
+				err = fmt.Errorf("containers not ready in pod: %v", pod.Name)
 				continue
 			}
 			klog.V(5).Infof("Pod %s %s is ready with %d container statuses, %d init container status, %d ephemeral statues",
 				pod.Name, pod.Namespace, len(pod.Status.ContainerStatuses), len(pod.Status.InitContainerStatuses), len(pod.Status.EphemeralContainerStatuses))
 			w.Mx.Lock()
-			err1 := w.fillInfo(pod, pod.Status.ContainerStatuses)
-			err2 := w.fillInfo(pod, pod.Status.InitContainerStatuses)
-			err3 := w.fillInfo(pod, pod.Status.EphemeralContainerStatuses)
+			for _, statuses := range [][]corev1.ContainerStatus{pod.Status.InitContainerStatuses, pod.Status.EphemeralContainerStatuses, pod.Status.ContainerStatuses} {
+				if err = w.fillInfo(pod, statuses); err != nil {
+					w.Mx.Unlock()
+					return err
+				}
+			}
```

Comment:
The previous implementation used to 
1. lock
2. try and fill Containers
3. try and fill InitContainers
4. try and fill EphemeralContainers
5. unlock

Whereas the current implementation 

1. lock
2. try and fill InitContainers
      - unlock and exit on error
3. try and fill EphemeralContainers
    - unlock and exit on error
4. try and fill Containers
    - unlock and exit on error
8. unlock on no errors

While I like the change of order, (Init, Emphermal, Containers), I think we shouldn't exit early. Rather preserve the old behaviour which is to continue to process the rest unless there is a good reason not to.